### PR TITLE
Unique specials mod

### PR DESCRIPTION
**Pull request by Perset**
#### Summary
Content "Creates a mod that renders some npc-driven locations unique"

#### Purpose of change
Eliminates the redundancy of several highly-unique, personalized locations with the same characters, missions, and features. Some of these locations have even been known to break each other if communicated with about the same missions. Saves map space and precious immersion.

#### Describe the solution
Adds the GLOBALLY_UNIQUE flag to the aforementioned locations.

#### Describe alternatives you've considered
Loosening the scope a little, the newer locations added by MNG are fairly generic, though I think they should at least be UNIQUE, if not globally.

#### Testing
Talk first, then test.

### CODE REVIEW

**Comment by TheShadowFerret** [comment]
Aren't some of these already globally unique(and if they aren't, should be in vanilla) 🤔

**Comment by Perset** [comment]
> 

not a one, as far as I can read. 

Caveat, though: I don't know how to read.

**Comment by TheShadowFerret** [comment]
Caveat, though: I don't know how to read.

Only the Locked LMOE Shelter & the teleported aircraft carrier have it, so apparently you can 😛

This should be a vanilla change.

**Comment by Perset** [comment]
> 

That was my last attempted PR, leashed to just Isherwood. Now that it's my mod, I get to expand things a little ;)

**Comment by nopenoperson** [comment]
should honestly be vanilla

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (Leland) [inline data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: The evac center is, for many new players, their first destination (found via the computer). Is making it globally unique liable to make it _extremely_ far from player spawn?

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: Does it not spawn on every overmap? How much do you need it to spawn? Also, calls for vanillization are not strictly related to this completely option _mod_...so I don't know what clueless new players have to do with anything?

Comment (Leland) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: Well, this PR may be a good candidate for moving to vanilla :) Performance issues are a healthy concern for mods, as well.

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: Ahhh...no, they've talked it down multiple times, it won't be added to vanilla. It's very important that the very special characters are seen...again and again.

Comment (Maleclypse) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: > Ahhh...no, they've talked it down multiple times, it won't be added to vanilla. It's very important that the very special characters are seen...again and again.

It won't be added to vanilla... until we are able to enforce spawning within a given radius of the player.

Comment (NetSysFire) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: I wonder if you could artificially crank up the occurances (e.g 40/100), so it will be generated towards the center (and hopefully spawn) and GLOBALLY_UNIQUE takes care that it will not be generated twice.

**Comment by PatrikLundell** [comment]
I find it this to be reasonable to have in a mod until such a time the game actually provides the support needed for it to be part of the base game. That will allow those who crave it to get it, while also finding out (and hopefully documenting) a lot of the issues resulting from locations being unique. The mod maintainer(s) may even help dealing with, or at least trying out solutions to, some of those issues.

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (Zireael07) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: @Leland [#61257](https://github.com/CleverRaven/Cataclysm-DDA/issues/61257?notification_referrer_id=NT_kwDOAEUNTLI0NDcyNzYxMDI4OjQ1MjUzODg&notifications_query=is%3Aunread#issuecomment-1281131036) is literally the latest of multiple discussions on having this in vanilla, and explains why it can't happen for now (TLDR: things like anthills/fungus/triffids can overrun the unique location even before the player gets there)

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:35]:
```suggestion
    "flags": [ "CLASSIC", "MAN_MADE" "WILDERNESS" ]
```
These locations are meant to spawn multiple times, rather than being unique. That's why I made them a bit generic, as you said.

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:40]:
```suggestion
    "flags": [ "CLASSIC", "MAN_MADE" ]
```
Same as above

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:45]:
```suggestion
    "flags": [ "CLASSIC", "MAN_MADE" ]
```
Here, too.

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]:
```suggestion
    "flags": [ "CLASSIC", "URBAN" ]
```
Here too.

**Review by MNG-cataclysm — CHANGES_REQUESTED** [review]

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]: Could they at least be UNIQUE, to keep them off of the same OM tile? That's what that flag does, right? The names might be generic but they still have identical railroaded text and missions.

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]: Yes...it does. As things stand, can two of the exact same trader appear on the same overmap chunk? Is that not what UNIQUE prevents?

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]: So I'll summarize my understanding with this question: do you need to see two occupied lumber mills in the same overmap chunk, doing and saying exactly the same thing and offering the same quests in the same order?

Comment (MNG-cataclysm) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]: UNIQUE prevents two (in this example, occupied lumbermills) from appearing in the same overmap chunk. However, multiple can spawn in the same world. GLOBALLY_UNIQUE only allows for one to spawn per world (you will only ever see it once in a game) If it is only UNIQUE, you can see multiple per game.

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]: Well, when nested mapgen has been taked such advantage of as to create that variety and mutability, that will be a fine day indeed! Instead, we are here as we are today. 

What satisfaction are you getting out of multiple identical traders on the same overmap? In a mod, not even suggested as active-by-default, that I'm trying to push?

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (anoobindisguise) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:10]: > It won't be added to vanilla... until we are able to enforce spawning within a given radius of the player. 
Something like forcing Exodii, Hub, Refugee, Artisans and Isherwoods to spawn in the middle 4 OMTs and prevent triffids/fungus/ants/etc from spawning in the middle 16 OMTs?

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json:50]: Multiple traders just seem logical to me. However, on reflection, I think I should try the nested mapgen idea for a bit more variety. On further reflection, I think the idea of just making them overmap-unique would work better.

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: Huh. Out of curiosity, how likely is a lumber mill to be occupied as opposed to unoccupied?

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: one out of every two lumber mills is a fortress, I think? Lumberjacks become the most numerous living humans left on Earth?

Comment (MNG-cataclysm) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: Sorry I didn't get back to you sooner.

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: I leave it to you to either insert occurrences that you think are appropriate either in the changes you're suggesting to this PR or directly to the main repo, if you like.

Promptness isn't an issue lol, no worries. Hopefully the content freeze doesn't descend before I can at least present this to the cabal.

Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: Um, also though, it says "outdated suggestions cannot be applied", dunno if there's something I'm supposed to do about that.

Comment (MNG-cataclysm) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: I wouldn't worry about that. I'll delete the suggested code and make a new one.

File: data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json:42]:
```suggestion
    "flags": [ "CLASSIC", "UNIQUE", "WILDERNESS" ],
    "occurrences": [ 15, 100 ]
```
The occurrence chances we discussed earlier.

File: data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json:48]:
```suggestion
    "flags": [ "CLASSIC", "UNIQUE" ],
    "occurrences": [ 15, 100 ]
```

File: data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json:54]:
```suggestion
    "flags": [ "CLASSIC", "UNIQUE" ],
    "occurrences": [ 15, 100 ]
```

File: data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json
Comment (MNG-cataclysm) [inline data/mods/Globally_Unique_Specials/overmap/overmap_special/specials.json:60]:
```suggestion
    "flags": [ "CLASSIC", "UNIQUE", "URBAN" ],
    "occurrences": [ 25, 100 ]
```

**Review by MNG-cataclysm — CHANGES_REQUESTED** [review]

**Review by MNG-cataclysm — APPROVED** [review]

File: data/mods/Globally Unique Specials/overmap/overmap_special/specials.json
Comment (Perset) [reply data/mods/Globally Unique Specials/overmap/overmap_special/specials.json]: OK...so, if copy-from doesn't have support for overmap specials, is there another elegant way I can do this mod? That will not require pasting the specials wholesale into the mod files?

**Comment by Perset** [comment]
> I seem to be getting consistent CTDs on my machine when using this mod locally. It'll cause a crash with the same error each time when attempting to find the refugee center using an evac shelter computer: devenv_KSH0p8ZGn2 devenv_eSnhPHe8BM
> 
> When attempting to place the overriden specials through the overmap editor, they appear to take up no space at all. Is it possible that the `copy-from` needs some additional explicit defs? Am I just missing something? The issues don't seem to happen when `GLOBALLY_UNIQUE` is added to the base defs manually, and I'd like to make sure it's just on my end.
> 
> Here's a game report, if it helps. I'm running some local changes on this version (as it's a working branch), but nothing affecting mapgen except for the mod.
> 
>     * OS: Windows
>       
>       * OS Version: 10.0.19044.2251 (21H2)
> 
>     * Game Version: 0.F-12376-g8f1620c07b-dirty [64-bit]
> 
>     * Graphics Version: Tiles
> 
>     * Game Language: System language []
> 
>     * Mods loaded: [
>       Dark Days Ahead [dda],
>       Blaze Industries [blazeindustries],
>       Bionic Slots [cbm_slots],
>       Disable NPC Needs [no_npc_food],
>       No Fungal Growth [no_fungal_growth],
>       No Rail Stations [No_Rail_Stations],
>       Globally Unique Specials [globally_unique_specials]
>       ]

AHHH! I'm super sorry, I just left this in the mud after I ran into (what I recall as) inheritance issues! Basically some functions that would make the mod easy to update don't work with some of what the mod makes changes to, and I thought "Damn that would suck to maintain" so I didn't bother to fine tune anything here on GH. I got what I wanted out of the mod, more or less, just playing around with stuff on my side, but again if changes are made to those files, the mod becomes useless until updated, so I just dropped it. If that makes any sense.

I'm sorry if I wasted any of your time or anybody else's.

But yeah a number of things had to be tweaked from what was left here.

**Pull request closed without merging**
